README.md
# Asyncio MOEX ISS API 

[![Build Status](https://travis-ci.org/WLM1ke/aiomoex.svg?branch=master)](https://travis-ci.org/WLM1ke/aiomoex)
[![Codacy Badge](https://api.codacy.com/project/badge/Coverage/363c10e1d85b404882326cf62b78f25c)](https://www.codacy.com/app/wlmike/aiomoex?utm_source=github.com&utm_medium=referral&utm_content=WLM1ke/aiomoex&utm_campaign=Badge_Coverage)
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/363c10e1d85b404882326cf62b78f25c)](https://www.codacy.com/app/wlmike/aiomoex?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=WLM1ke/aiomoex&amp;utm_campaign=Badge_Grade)

Реализация на основе asyncio части  запросов к [MOEX Informational & Statistical Server](https://www.moex.com/a2193)

## Основные возможности
Реализовано несколько функций-запросов информации о торгуемых акциях и их исторических котировках, результаты которых
напрямую конвертируются в pandas.DataFrame.

Работа функций базируется на универсальном клиенте, позволяющем осуществлять произвольные запросы к MOEX ISS, поэтому
перечень доступных функций-запросов может быть легко расширен. При необходимости добавления функций воспользуйтесь
[Issues](https://github.com/WLM1ke/aiomoex/issues>) на GitHub с указанием ссылки на описание запроса:
- Полный перечень [запросов](https://iss.moex.com/iss/reference/) запросов к MOEX ISS
- Официальное [Руководство разработчика](https://fs.moex.com/files/6523) с дополнительной информацией

## Начало работы
### Установка 
```
pip install aiomoex
```

### Пример использования реализованных запросов
История котировок SNGSP в режиме TQBR
```
import asyncio

import aiomoex
import pandas as pd


async def main():
    async with aiomoex.ISSClientSession():
        data = await aiomoex.get_board_history('SNGSP')
        df = pd.DataFrame(data)
        df.set_index('TRADEDATE', inplace=True)
        print(df.head(), '\n')
        print(df.tail(), '\n')
        df.info()


asyncio.run(main())
```

```
            CLOSE    VOLUME
TRADEDATE                  
2014-06-09  27.48  12674200
2014-06-10  27.55  14035900
2014-06-11  28.15  27208800
2014-06-16  28.27  68059900
2014-06-17  28.20  22101600 

             CLOSE    VOLUME
TRADEDATE                   
2018-11-16  37.860   9660800
2018-11-19  37.315  28765600
2018-11-20  36.790  19853500
2018-11-21  36.930  14583000
2018-11-22  37.480   9656600 

<class 'pandas.core.frame.DataFrame'>
Index: 1125 entries, 2014-06-09 to 2018-11-22
Data columns (total 2 columns):
CLOSE     1125 non-null float64
VOLUME    1125 non-null int64
dtypes: float64(1), int64(1)
memory usage: 26.4+ KB
```

### Пример реализации запроса с помощью клиента
Перечень акций, торгующихся в режиме TQBR - [описание запроса](https://iss.moex.com/iss/reference/32)
```
import asyncio

import aiomoex
import pandas as pd


async def main():
    request_url = 'https://iss.moex.com/iss/engines/stock/markets/shares/boards/TQBR/securities.json'
    arguments = {'securities.columns': 'SECID,REGNUMBER,LOTSIZE,SHORTNAME'}
    iss = aiomoex.ISSClient(request_url, arguments)
    iss.start_session()
    data = await iss.get()
    df = pd.DataFrame(data['securities'])
    df.set_index('SECID', inplace=True)
    print(df.head(), '\n')
    print(df.tail(), '\n')
    df.info()
    await iss.close_session()


asyncio.run(main())
```
```
       LOTSIZE     REGNUMBER   SHORTNAME
SECID                                   
ABRD        10  1-02-12500-A  АбрауДюрсо
AFKS       100  1-05-01669-A  Система ао
AFLT        10  1-01-00010-A    Аэрофлот
AGRO         1          None    AGRO-гдр
AKRN         1  1-03-00207-A       Акрон 

       LOTSIZE     REGNUMBER  SHORTNAME
SECID                                  
YRSBP      100  2-01-50099-A  ТНСэнЯр-п
ZILL         1  1-02-00036-A     ЗИЛ ао
ZMZN        10  1-01-00230-A     ЗМЗ-ао
ZMZNP      100  2-01-00230-A     ЗМЗ-ап
ZVEZ      1000  1-01-00169-D  ЗВЕЗДА ао 

<class 'pandas.core.frame.DataFrame'>
Index: 277 entries, ABRD to ZVEZ
Data columns (total 3 columns):
LOTSIZE      277 non-null int64
REGNUMBER    268 non-null object
SHORTNAME    277 non-null object
dtypes: int64(1), object(2)
memory usage: 8.7+ KB
```



docs/change_log.rst
Список изменений
================

2.1.1 (2023-11-27)
------------------
* Добавлен py.typed

2.1.0 (2023-11-27)
------------------
* Минимальная версия Python 3.10
* Обновлены зависимости
* Переход на rye, ruff и pyright

2.0.0 (2020-09-08)
------------------
* Изменены все сигнатуры - сессия aiohttp передается в явном виде
* Минимальная версия Python 3.8
* Полная анотация типов и проверкой `MyPy <https://mypy.readthedocs.io/en/stable/>`_
* Рефакторинг с учетом требований `wemake-python-styleguide <https://wemake-python-stylegui.de/en/latest/>`_

1.2.2 (2018-12-07)
------------------
* Исправлена ошибка при обработке некоторых пустых ответов MOEX ISS

1.2.1 (2018-12-01)
------------------
* Добавлена дополнительная проверка версии при установке

1.2.0 (2018-11-28)
------------------
* Добавлена функция получения интервала дат, доступных в истории для рынка по заданному режиму торгов get_board_dates()

1.1.1 (2018-11-27)
------------------
* Для исторических котировок в перечень загружаемых по умолчанию колонок добавлены режим торгов и объем в рублях

1.1.0 (2018-11-27)
------------------
* Добавлена функция получения справочника разделов MOEX ISS get_reference()
* Добавлены функции для получения исторических данных по свечкам get_market_candle_borders(), get_market_candles(), get_board_candle_borders() и get_board_candles()
* Добавлена группировка функций в документации
* Удалена функция get_index_history() - эквивалентна вызову get_board_history('MCFTRR', start, end, columns, 'RTSI', 'index')

1.0.1 (2018-11-25)
------------------
* Добавлено сравнение скорости

1.0.0 (2018-11-24)
------------------
* Начальная версия


aiomoex/requests.py
"""Реализация части запросов к MOEX ISS

При необходимости могут быть дополнены:
    Полный перечень запросов https://iss.moex.com/iss/reference/
    Дополнительное описание https://fs.moex.com/files/6523
"""
from typing import Iterable, Optional

import aiohttp

__all__ = [
    "get_reference",
    "find_securities",
    "get_market_candle_borders",
    "get_board_candle_borders",
    "get_market_candles",
    "get_board_candles",
    "get_board_dates",
    "get_board_securities",
    "get_market_history",
    "get_board_history",
]

from aiomoex import client


def _make_query(
    *,
    q: Optional[str] = None,
    interval: Optional[int] = None,
    start: Optional[str] = None,
    end: Optional[str] = None,
    table: Optional[str] = None,
    columns: Optional[Iterable[str]] = None,
) -> client.WebQuery:
    """Формирует дополнительные параметры запроса к MOEX ISS

    В случае False значений не добавляются в запрос

    :param q:
        Строка с частью характеристик бумаги для поиска
    :param interval:
        Размер свечки
    :param start:
        Начальная дата котировок
    :param end:
        Конечная дата котировок
    :param table:
        Таблица, которую нужно загрузить (для запросов, предполагающих наличие нескольких таблиц)
    :param columns:
        Кортеж столбцов, которые нужно загрузить

    :return:
        Словарь с дополнительными параметрами запроса
    """
    query: client.WebQuery = dict()
    if q:
        query["q"] = q
    if interval:
        query["interval"] = interval
    if start:
        query["from"] = start
    if end:
        query["till"] = end
    if table:
        query["iss.only"] = f"{table},history.cursor"
    if columns:
        query[f"{table}.columns"] = ",".join(columns)
    return query


def _get_table(data: client.TablesDict, table_name: str) -> client.Table:
    """Извлекает конкретную таблицу из данных"""
    try:
        table = data[table_name]
    except KeyError:
        raise client.ISSMoexError(f"Отсутствует таблица {table_name} в данных")
    return table


async def _get_short_data(
    session: aiohttp.ClientSession, url: str, table_name: str, query: Optional[client.WebQuery] = None,
) -> client.Table:
    """Получить данные для запроса с выдачей всей информации за раз

    :param session:
        Сессия http соединения.
    :param url:
        URL запроса
    :param query:
        Дополнительные параметры запроса - None, если нет параметров
    :param table_name:
        Таблица, которую нужно выбрать

    :return:
        Конкретная таблица из запроса
    """
    iss = client.ISSClient(session, url, query)
    data = await iss.get()
    return _get_table(data, table_name)


async def _get_long_data(
    session: aiohttp.ClientSession, url: str, table_name: str, query: Optional[client.WebQuery] = None,
) -> client.Table:
    """Получить данные для запроса, в котором информация выдается несколькими блоками

    :param session:
        Сессия http соединения.
    :param url:
        URL запроса
    :param query:
        Дополнительные параметры запроса - None, если нет параметров
    :param table_name:
        Таблица, которую нужно выбрать

    :return:
        Конкретная таблица из запроса
    """
    iss = client.ISSClient(session, url, query)
    data = await iss.get_all()
    return _get_table(data, table_name)


async def get_reference(session: aiohttp.ClientSession, placeholder: str = "boards") -> client.Table:
    """Получить перечень доступных значений плейсхолдера в адресе запроса

    Например в описание запроса https://iss.moex.com/iss/reference/32 присутствует следующий адрес
    /iss/engines/[engine]/markets/[market]/boards/[board]/securities с плейсхолдерами engines, markets и boards

    Для работы требуется открытая ISSClientSession

    Описание запроса - https://iss.moex.com/iss/reference/28

    :param session:
        Сессия http соединения.
    :param placeholder:
        Наименование плейсхолдера в адресе запроса: engines, markets, boards, boardgroups, durations, securitytypes,
        securitygroups, securitycollections

    :return:
        Список словарей, которые напрямую конвертируется в pandas.DataFrame
    """
    url = "https://iss.moex.com/iss/index.json"
    return await _get_short_data(session, url, placeholder)


async def find_securities(
    session: aiohttp.ClientSession,
    string: str,
    columns: Optional[Iterable[str]] = ("secid", "regnumber"),
) -> client.Table:
    """Найти инструменты по части Кода, Названию, ISIN, Идентификатору Эмитента, Номеру гос.регистрации

    Один из вариантов использования - по регистрационному номеру узнать предыдущие тикеры эмитента, и с помощью
    нескольких запросов об истории котировок собрать длинную историю с использованием всех предыдущих тикеров

    Для работы требуется открытая ISSClientSession

    Описание запроса - https://iss.moex.com/iss/reference/32

    :param session:
        Сессия http соединения.
    :param string:
        Часть Кода, Названия, ISIN, Идентификатора Эмитента, Номера гос.регистрации
    :param columns:
        Кортеж столбцов, которые нужно загрузить - по умолчанию тикер и номер государственно регистрации.
        Если пустой или None, то загружаются все столбцы

    :return: Список словарей, которые напрямую конвертируется в pandas.DataFrame
    """
    url = "https://iss.moex.com/iss/securities.json"
    table = "securities"
    query = _make_query(q=string, table=table, columns=columns)
    return await _get_short_data(session, url, table, query)


async def get_market_candle_borders(
    session: aiohttp.ClientSession, security: str, market: str = "shares", engine: str = "stock",
) -> client.Table:
    """Получить таблицу интервалов доступных дат для свечей различного размера на рынке для всех режимов торгов

    Для работы требуется открытая ISSClientSession

    Описание запроса - https://iss.moex.com/iss/reference/156

    :param session:
        Сессия http соединения.
    :param security:
        Тикер ценной бумаги
    :param market:
        Рынок - по умолчанию акции
    :param engine:
        Движок - по умолчанию акции

    :return:
        Список словарей, которые напрямую конвертируется в pandas.DataFrame
    """
    url = f"https://iss.moex.com/iss/engines/{engine}/markets/{market}/securities/{security}/candleborders.json"
    table = "borders"
    return await _get_short_data(session, url, table)


async def get_board_candle_borders(
    session: aiohttp.ClientSession,
    security: str,
    board: str = "TQBR",
    market: str = "shares",
    engine: str = "stock",
) -> client.Table:
    """Получить таблицу интервалов доступных дат для свечей различного размера в указанном режиме торгов

    Для работы требуется открытая ISSClientSession

    Описание запроса - https://iss.moex.com/iss/reference/48

    :param session:
        Сессия http соединения.
    :param security:
        Тикер ценной бумаги
    :param board:
        Режим торгов - по умолчанию основной режим торгов T+2
    :param market:
        Рынок - по умолчанию акции
    :param engine:
        Движок - по умолчанию акции

    :return:
        Список словарей, которые напрямую конвертируется в pandas.DataFrame
    """
    url = (
        f"https://iss.moex.com/iss/engines/{engine}/markets/{market}/"
        f"boards/{board}/securities/{security}/candleborders.json"
    )
    table = "borders"
    return await _get_short_data(session, url, table)


async def get_market_candles(
    session: aiohttp.ClientSession,
    security: str,
    interval: int = 24,
    start: Optional[str] = None,
    end: Optional[str] = None,
    market: str = "shares",
    engine: str = "stock",
) -> client.Table:
    """Получить свечи в формате HLOCV указанного инструмента на рынке для основного режима торгов за интервал дат

    Если торговля идет в нескольких основных режимах, то на один интервал времени может быть выдано несколько свечек -
    по свечке на каждый режим. Предположительно такая ситуация может произойти для свечек длиннее 1 дня

    Для работы требуется открытая ISSClientSession

    Описание запроса - https://iss.moex.com/iss/reference/155

    :param session:
        Сессия http соединения.
    :param security:
        Тикер ценной бумаги
    :param interval:
        Размер свечки - целое число 1 (1 минута), 10 (10 минут), 60 (1 час), 24 (1 день), 7 (1 неделя), 31 (1 месяц) или
        4 (1 квартал). По умолчанию дневные данные
    :param start:
        Дата вида ГГГГ-ММ-ДД. При отсутствии данные будут загружены с начала истории
    :param end:
        Дата вида ГГГГ-ММ-ДД. При отсутствии данные будут загружены до конца истории
    :param market:
        Рынок - по умолчанию акции
    :param engine:
        Движок - по умолчанию акции

    :return:
        Список словарей, которые напрямую конвертируется в pandas.DataFrame
    """
    url = (
        f"https://iss.moex.com/iss/engines/{engine}/markets/{market}/securities/{security}/candles.json"
    )
    table = "candles"
    query = _make_query(interval=interval, start=start, end=end)
    return await _get_long_data(session, url, table, query)


async def get_board_candles(
    session: aiohttp.ClientSession,
    security: str,
    interval: int = 24,
    start: Optional[str] = None,
    end: Optional[str] = None,
    board: str = "TQBR",
    market: str = "shares",
    engine: str = "stock",
) -> client.Table:
    """Получить свечи в формате HLOCV указанного инструмента в указанном режиме торгов за интервал дат

    Для работы требуется открытая ISSClientSession

    Описание запроса - https://iss.moex.com/iss/reference/46

    :param session:
        Сессия http соединения.
    :param security:
        Тикер ценной бумаги
    :param interval:
        Размер свечки - целое число 1 (1 минута), 10 (10 минут), 60 (1 час), 24 (1 день), 7 (1 неделя), 31 (1 месяц) или
        4 (1 квартал). По умолчанию дневные данные
    :param start:
        Дата вида ГГГГ-ММ-ДД. При отсутствии данные будут загружены с начала истории
    :param end:
        Дата вида ГГГГ-ММ-ДД. При отсутствии данные будут загружены до конца истории
    :param board:
        Режим торгов - по умолчанию основной режим торгов T+2
    :param market:
        Рынок - по умолчанию акции
    :param engine:
        Движок - по умолчанию акции

    :return:
        Список словарей, которые напрямую конвертируется в pandas.DataFrame
    """
    url = (
        f"https://iss.moex.com/iss/engines/{engine}/markets/{market}/"
        f"boards/{board}/securities/{security}/candles.json"
    )
    table = "candles"
    query = _make_query(interval=interval, start=start, end=end)
    return await _get_long_data(session, url, table, query)


async def get_board_dates(
    session: aiohttp.ClientSession, board: str = "TQBR", market: str = "shares", engine: str = "stock",
) -> client.Table:
    """Получить интервал дат, доступных в истории для рынка по заданному режиму торгов

    Для работы требуется открытая ISSClientSession

    Описание запроса - https://iss.moex.com/iss/reference/26

    :param session:
        Сессия http соединения.
    :param board:
        Режим торгов - по умолчанию основной режим торгов T+2
    :param market:
        Рынок - по умолчанию акции
    :param engine:
        Движок - по умолчанию акции

    :return:
        Список из одного элемента - словаря с ключами 'from' и 'till'
    """
    url = f"https://iss.moex.com/iss/history/engines/{engine}/markets/{market}/boards/{board}/dates.json"
    table = "dates"
    return await _get_short_data(session, url, table)


async def get_board_securities(
    session: aiohttp.ClientSession,
    table: str = "securities",
    columns: Optional[Iterable[str]] = ("SECID", "REGNUMBER", "LOTSIZE", "SHORTNAME"),
    board: str = "TQBR",
    market: str = "shares",
    engine: str = "stock",
) -> client.Table:
    """Получить таблицу инструментов по режиму торгов со вспомогательной информацией

    Для работы требуется открытая ISSClientSession

    Описание запроса - https://iss.moex.com/iss/reference/32

    :param session:
        Сессия http соединения.
    :param table:
        Таблица с данными, которую нужно вернуть: securities - справочник торгуемых ценных бумаг, marketdata -
        данные с результатами торгов текущего дня
    :param columns:
        Кортеж столбцов, которые нужно загрузить - по умолчанию тикер, номер государственно регистрации,
        размер лота и краткое название. Если пустой или None, то загружаются все столбцы
    :param board:
        Режим торгов - по умолчанию основной режим торгов T+2
    :param market:
        Рынок - по умолчанию акции
    :param engine:
        Движок - по умолчанию акции

    :return:
        Список словарей, которые напрямую конвертируется в pandas.DataFrame
    """
    url = f"https://iss.moex.com/iss/engines/{engine}/markets/{market}/boards/{board}/securities.json"
    query = _make_query(table=table, columns=columns)
    return await _get_short_data(session, url, table, query)


async def get_market_history(
    session: aiohttp.ClientSession,
    security: str,
    start: Optional[str] = None,
    end: Optional[str] = None,
    columns: Optional[Iterable[str]] = ("BOARDID", "TRADEDATE", "CLOSE", "VOLUME", "VALUE"),
    market: str = "shares",
    engine: str = "stock",
) -> client.Table:
    """Получить историю по одной бумаге на рынке для всех режимов торгов за интервал дат

    На одну дату может приходиться несколько значений, если торги шли в нескольких режимах

    Для работы требуется открытая ISSClientSession

    Описание запроса - https://iss.moex.com/iss/reference/63

    :param session:
        Сессия http соединения.
    :param security:
        Тикер ценной бумаги
    :param start:
        Дата вида ГГГГ-ММ-ДД. При отсутствии данные будут загружены с начала истории
    :param end:
        Дата вида ГГГГ-ММ-ДД. При отсутствии данные будут загружены до конца истории
    :param columns:
        Кортеж столбцов, которые нужно загрузить - по умолчанию режим торгов, дата торгов, цена закрытия и объем в
        штуках и стоимости. Если пустой или None, то загружаются все столбцы
    :param market:
        Рынок - по умолчанию акции
    :param engine:
        Движок - по умолчанию акции

    :return:
        Список словарей, которые напрямую конвертируется в pandas.DataFrame
    """
    url = (
        f"https://iss.moex.com/iss/history/engines/{engine}/markets/{market}/securities/{security}.json"
    )
    table = "history"
    query = _make_query(start=start, end=end, table=table, columns=columns)
    return await _get_long_data(session, url, table, query)


async def get_board_history(
    session: aiohttp.ClientSession,
    security: str,
    start: Optional[str] = None,
    end: Optional[str] = None,
    columns: Optional[Iterable[str]] = ("BOARDID", "TRADEDATE", "CLOSE", "VOLUME", "VALUE"),
    board: str = "TQBR",
    market: str = "shares",
    engine: str = "stock",
) -> client.Table:
    """Получить историю торгов для указанной бумаги в указанном режиме торгов за указанный интервал дат

    Для работы требуется открытая ISSClientSession

    Описание запроса - https://iss.moex.com/iss/reference/65

    :param session:
        Сессия http соединения.
    :param security:
        Тикер ценной бумаги
    :param start:
        Дата вида ГГГГ-ММ-ДД. При отсутствии данные будут загружены с начала истории
    :param end:
        Дата вида ГГГГ-ММ-ДД. При отсутствии данные будут загружены до конца истории
    :param columns:
        Кортеж столбцов, которые нужно загрузить - по умолчанию режим торгов, дата торгов, цена закрытия и объем в
        штуках и стоимости. Если пустой или None, то загружаются все столбцы
    :param board:
        Режим торгов - по умолчанию основной режим торгов T+2
    :param market:
        Рынок - по умолчанию акции
    :param engine:
        Движок - по умолчанию акции

    :return:
        Список словарей, которые напрямую конвертируется в pandas.DataFrame
    """
    url = (
        f"https://iss.moex.com/iss/history/engines/{engine}/markets/{market}/"
        f"boards/{board}/securities/{security}.json"
    )
    table = "history"
    query = _make_query(start=start, end=end, table=table, columns=columns)
    return await _get_long_data(session, url, table, query)
